src/platform/test.rs

```diff
@@ -653,3 +770,103 @@ mod sync_test {
+    let (sub_tx, sub_rx) = platform::channel().unwrap();
+
+    // send the other process the tx side, so it can send us the channels
+    super_tx.send(&[], vec![OsIpcChannel::Sender(sub_tx)], vec![]).unwrap();
+
+    // get two_rx from the other process
+    let (_, mut received_channels, _) = sub_rx.recv().unwrap();
+    assert_eq!(received_channels.len(), 1);
+    let two_rx = received_channels.pop().unwrap().to_receiver();
+
+    // get one_rx from two_rx's buffer
+    let (_, mut received_channels, _) = two_rx.recv().unwrap();
+    assert_eq!(received_channels.len(), 1);
+    let one_rx = received_channels.pop().unwrap().to_receiver();
+
+    // get a cookie from one_rx
+    let (mut data, _, _) = one_rx.recv().unwrap();
+    data.truncate(cookie.len());
+    assert_eq!(&data[..], cookie);
+
+    // finally, send a cookie back
+    super_tx.send(&data, vec![], vec![]).unwrap();
+
+    // terminate
+    unsafe { libc::exit(0); }
+}
+
+#[cfg(not(any(feature = "force-inprocess", target_os = "android")))]
+#[cfg_attr(target_os = "windows", should_panic(expected = "ChannelClosed"))]
+#[test]
+fn cross_process_two_step_transfer_spawn() {
+    let cookie: &[u8] = b"cookie";
+
+    // create channel 1
+    let (one_tx, one_rx) = platform::channel().unwrap();
+    // put data in channel 1's pipe
+    one_tx.send(cookie, vec![], vec![]).unwrap();
+
+    // create channel 2
+    let (two_tx, two_rx) = platform::channel().unwrap();
+    // put channel 1's rx channel in channel 2's pipe
+    two_tx.send(&[], vec![OsIpcChannel::Receiver(one_rx)], vec![]).unwrap();
+
+    // create a one-shot server, and spawn another process
+    let (server, name) = OsIpcOneShotServer::new().unwrap();
+    let mut child_pid = Command::new(env::current_exe().unwrap())
+        .arg("--ignored")
+        .arg("cross_process_two_step_transfer_server")
+        .arg(format!("channel_name:{}", name))
+        .stdin(Stdio::null())
+        .stdout(Stdio::null())
+        .stderr(Stdio::null())
+        .spawn()
+        .expect("failed to execute server process");
+
+    // The other process will have sent us a transmit channel in received channels
+    let (super_rx, _, mut received_channels, _) = server.accept().unwrap();
+    assert_eq!(received_channels.len(), 1);
+    let sub_tx = received_channels.pop().unwrap().to_sender();
+
+    // We use the transmit portion to send channel 2's rx to the other process
+    sub_tx.send(&[], vec![OsIpcChannel::Receiver(two_rx)], vec![]).unwrap();
+
+    // Then we wait for the cookie to make its way back to us
+    let (mut received_data, received_channels, received_shared_memory_regions) =
+        super_rx.recv().unwrap();
+    let child_exit_code = child_pid.wait().expect("failed to wait on child");
+    assert!(child_exit_code.success());
+    received_data.truncate(cookie.len());
+    assert_eq!((&received_data[..], received_channels, received_shared_memory_regions),
+               (cookie, vec![], vec![]));
+}
```

Comment:
I realised that there is another, more interesting case that this test doesn't cover yet (I *knew* there is something missing... ;-) ): sending data or other channels over an outer channel while the receive end of that outer channel is already in transfer.

However, AIUI this should probably just work even with your current implementation?...